hibernate-core/src/test/java/org/hibernate/test/idgen/foreign/ForeignGeneratorResourceLocalTest.java

```diff
@@ -36,275 +34,366 @@
-		}
-
-		public void setContractId(Long contractId) {
-			id.setContractId( contractId );
-		}
-	}
+@TestForIssue(jiraKey = "HHH-12738")
+public class ForeignGeneratorResourceLocalTest extends BaseEntityManagerFunctionalTestCase
+{
+
+    @Override
+    protected Class<?>[] getAnnotatedClasses()
+    {
+        return new Class[]{
+            Contract.class,
+            Customer.class,
+            CustomerContractRelation.class
+        };
+    }
+
+    @Test
+    public void addRelationImplicitFlush() throws Exception
+    {
+
+        Long contractId = doInJPA(this::entityManagerFactory, entityManager ->
+        {
+            Contract contract = new Contract();
+
+            entityManager.persist(contract);
+            return contract.getId();
+        });
+
+        Long customerId = doInJPA(this::entityManagerFactory, entityManager ->
+        {
+            Customer customer = new Customer();
+
+            entityManager.persist(customer);
+            return customer.getId();
+        });
+
+        doInJPA(this::entityManagerFactory, entityManager ->
+        {
+
+            Customer customer = entityManager.createQuery(
+                "SELECT c " +
+                    "FROM Customer c " +
+                    " LEFT JOIN FETCH c.contractRelations " +
+                    " WHERE c.id = :customerId",
+                Customer.class)
+                .setParameter("customerId", customerId)
+                .getSingleResult();
+
+            CustomerContractRelation relation = new CustomerContractRelation();
+            relation.setContractId(contractId);
+            customer.addContractRelation(relation);
+        });
+    }
+
+    @Test
+    public void addRelationImplicitFlushClosedEM() throws Exception
+    {
+
+        Long contractId = doInJPA(this::entityManagerFactory, entityManager ->
+        {
+            Contract contract = new Contract();
+
+            entityManager.persist(contract);
+            return contract.getId();
+        });
+
+        Long customerId = doInJPA(this::entityManagerFactory, entityManager ->
+        {
+            Customer customer = new Customer();
+
+            entityManager.persist(customer);
+            return customer.getId();
+        });
+
+        doInJPA(this::entityManagerFactory, entityManager ->
+        {
+
+            Customer customer = entityManager.createQuery(
+                "SELECT c " +
+                    "FROM Customer c " +
+                    " LEFT JOIN FETCH c.contractRelations " +
+                    " WHERE c.id = :customerId",
+                Customer.class)
+                .setParameter("customerId", customerId)
+                .getSingleResult();
+
+            CustomerContractRelation relation = new CustomerContractRelation();
+            relation.setContractId(contractId);
+            customer.addContractRelation(relation);
+
+            // simulating spring JtaTransactionManager.triggerBeforeCompletion()
+            entityManager.close();
```

Comment:
That's not a valid workflow. Do you want to span a Tx over multiple EntityManagers?

Reply:
Hi Vlad! Thx for your quick reply. I am totally aware about that and I would not close the session as well. However, this is how Spring does the integration. 

Well, within the project I am working in, we have multiple entity managers and spawning Tx over them using Bitronix as XA transaction manager. This was the setup I originaly wanted to provid by my test case including the spring and bitronix dependencies. 

As a workaround for this issue I am able to manually flush the entity manager with the business code and everything is going well. Only if a left this work to the environment (spring), I run into this problems. However, the workaround seems to me not very practicable at all.

I am not sure how to proceed. As already mentioned, spring contributors are quite sure, that closing the entity manager before starting the commit phase should be no problem at all. And of cause, it seems to be no problem except the situation in the ForeignGenerator. 

Thx for your time and patience!

Reply:
@mjesseni I managed to replicate it, so I reopened it. I'll propose a fix today.

Reply:
@vladmihalcea thx a lot!